### ISSUE TITLE
Some ideas about the future of DOM

### ISSUE THREAD

**Issue by milankinen** [motorcyclejs/motorcyclejs#35]
Cheers! This issue contains brainstorming and some ideas how the "future of DOM" would look like in my eyes. These are not direct feature requests, instead I'm hoping that these topics raise some discussion that (in the end) leads to decisions useful for the entire community.

I've already discussed about this in Cycle community but Cycle's ideology and these ideas seem contradictory, hence I'm continuing here in the case that Motorcycle community shares at least some of them. 😄 

---

### Transposition

Cycle had transposition (or "embedding observables into vdom") but it was removed lately. I'm not so familiar with the reasons but in my opinion that is one of the best features one can have in vdom, and proposing that it'll be included in `@motorcycle/dom` because:
* It enables **blazing fast** DOM updates; when something changes in vdom, only the changed vnode can be diffed and patched in O(1) time
* It reduces boilerplate and redundancy because there is no need to combine observables before using them
* It's backwards compatible and opt-in; you can still use `O.combine` with it without having any negative performance / API usage impact

Consider `BmiCalculator` from [cycle-examples](https://github.com/cyclejs/cyclejs/blob/master/examples/bmi-nested/src/BmiCalculator.js):
```js
function view(bmi$, weightSliderDOM, heightSliderDOM) {
  return xs.combine(bmi$, weightSliderDOM, heightSliderDOM)
    .map(([bmi, weightVTree, heightVTree]) =>
      div([
        weightVTree,
        heightVTree,
        h2(`BMI is ${bmi}`)
      ])
    );
}
```
Could be written as:
```js
function view(bmi$, weightSliderDOM, heightSliderDOM) {
  return div([
        weightVTree,
        heightVTree,
        h2(bmi$.map(bmi => `BMI is ${bmi}`))
      ])
}
```

You may also notice an interesting extra feature: `view` function doesn't need to know whether DOM params are observables or not. This makes e.g. unit testing easier and makes components more reusable; consider this:
```js
function LabeledInput({title, value}) {
  return h("label", [
    title,
    h("input", {type: "text", value})
  ])
} 
```
Now the labeled input view doesn't need to know about the exact details of `title` and `value`, hence it can be called in many ways:
```js
const inputA = LabeledInput({title: "Tsers", value: state.select("tsers")})
const inputB = LabeledInput({title: state.map(({foo}) => foo ? "Foo" : "Bar"), value: state.select("bar")})
const inputC = LabeledInput({title: "Const", value: "123"})
```

#### About type safety and "magic"

One might argue that embedded observables are not type safe or some "implicit magic". I don't see it that way - it just the contract of vnode. How about `O.combine` (`combine :: [Observable<A>] => Observable<[A]>`), is that any more magic? Nope, it's just a way how combine works.

When expressed with types, the signature of `h` is something like this:
```
type Prop = string | num
type Child = string | VNode
type Children = Child | [Child]
type VNode = {value: Prop?, ..., children: Children}
h :: (string, {value: Prop?, ...}?, Children?) => VNode
```

With transposition, the signature is a bit more complicated but still strongly typed:
```
type Prop = string | num | 
type ObsProp = Prop | Observable<Prop>
type Child = string | VNode
type ObsChild = Child | Observable<Child>
type Children = ObsChild | [ObsChild] | Observable<[Child]>
type VNode = {value: ObsProp?, ..., children: Children}
h :: (string, {value: ObsProp?, ...}?, Children?) => VNode
```
 

### Event subscriptions

For a long time in (motor)cycle apps, DOM event streams have been obtained by using `DOMSource`, e.g.
```js
const click$ = DOM.select(".foo").events("click")
```

This approach favors the "traditional" cycle state management where model depends on intent and it has been beneficial to get events "first":
```js
// intent
const plus$ = DOM.select(".inc").events("click").constant(+1)
const minus$ = DOM.select(".dec").events("click").constant(-1)

// model
const sum$ = O.merge(plus$, minus$).scan(R.add, 0)

// view
const vdom$ = sum$.map(sum => ...)
```

This works pretty well in toy apps, but apps with more complicated state management (real apps IMO) this approach tends to enforce cumbersome and complex code. This has been (finally!) understood in Cycle community, thus libraries like `onionify` have been developed.

An another problem in this approach is that there must be a way to isolate events targeted to two or more sibling components. Because the events are coming directly from DOM the only way to distinguish siblings is to make 2-staged isolation: (1) assign some identity to component's root DOM node and (2) filter emitted events based on that identity.

In cycle, `isolate` handles this operation:
```js
const IsolatedComp = isolate(Comp, id)
```

Note that in order to preserve referential transparency ("purity"), `id` must given to the isolation. If I've understood correctly, there won't be any `isolate` in Motorcycle. This means that the user must perform both of those tasks and define id manually every time when isolation is needed:
```js
const a = Comp({...sources, DOM: DOM.select(".a")})
const b = Comp({...sources, DOM: DOM.select(".b")})

const vdom = h("div", [
  h("div.a", a.DOM),
  h("div.b", b.DOM)
])
```
This however, introduces (IMHO) unnecessary boilerplate and requires extra DOM nodes (or even more boilerplate with `a.DOM.map(withClass)`) in order to ensure isolation. In addition, events may still "leak" from child component to parent if the parent component subscribes to events by using selector matching child component nodes. This is also more error prone because class selectors must be defined in multiple places.

#### Monadic event subscription

I see cycle applications just as monadic functions projecting state (over time) into view (over time) and a bunch of control events/actions `Main :: Behaviour<State> => [Behaviour<View>, EventStream<Action>]`. *How about if we modeled DOM and it's event in a similar way? How about if events could be subscribed directly from `vnode` instead of `DOMSource`?*

```js
// model
function main({onion: state}) {
  // view
  const vdom = h("div", [
    h("h1", state.value.map(name => `Hello ${name}!`),
    h("button.send", "Send me!"),
    h("input.name", {type: "text", value: state.value})
  ])
  // intent
  const click$ = vdom.events(".send", "click")
  const type$ = vdom.events(".name", "input")
  return {
    DOM: vdom,
    onion: O.merge(
      click$.map(() => () => ""),
      type$.map(e => () => e.target.value)
    )
  } 
}
```

This proposal of course requires that the "official" state management supports circular dependency by default (like `onionify` does) because it changes the dependency order from `I -> M` to `M -> I`. 

But what benefits do we get by doing this this? As you might have noticed, there is no dependency to `DOM` when defining events, hence we don't need to perform any filtering in our parent component. And because we don't need to do filtering, we don't need to encode any filtering info into our DOM nodes. Parent nodes become extremely simple:
```js
const a = Comp(sources)
const b = Comp(sources)

const vdom = h("div", [ a.DOM, b.DOM ])
```

Of course there is still a situation where events "leak" from child component to parent. I see two possible alternatives to tackle this:

* Introduce an optional `isolated` function that can be used in parent component if the isolation must be ensured
```js
const a = Comp(sources)
const b = Comp(sources)

const vdom = h("div", [ isolated(a.DOM), isolated(b.DOM) ])
```

* Do not give any option to use selector when defining events; events may only be defined directly by using the vnode associated with them
```js
 function main({onion: state}) {
  let send, name
  const vdom = h("div", [
    h("h1", state.value.map(name => `Hello ${name}!`),
    send = h("button.send", "Send me!"),
    name = h("input.name", {type: "text", value: state.value})
  ])
  // intent
  const click$ = send.events("click")
  const type$ = name.events("input")
  return {
    DOM: vdom,
    onion: O.merge(
      click$.map(() => () => ""),
      type$.map(e => () => e.target.value)
    )
  } 
}
```

---

That was my 2 cents. Looking forward to some discussion about the topic. 😄

**Comment by TylorS** [comment]
Not able to read this right now, but have marked it as a discussion for us to have. Whatever the outcome, thanks for opening up this discussion.

**Comment by TylorS** [comment]
@milankinen There is a lot of stuff here, and I really appreciate the time spent to put this stuff together! 

@Frikki and I would love to have a video chat at some point to have a real conversation of these things to really get a feel for all sides of these topics, this weekend would be perfect, if that works for you.

**Comment by Frikki** [comment]
Yes, @milankinen . I understand what you are suggesting. However, we’ve had some hands on experience with your suggestions, wherefore it’ll likely be good to talk them over.

Particularly, transposition actually slowed it down, instead of, as we might think, speed it up. Perhaps we were doing it wrong and you might have some ideas on that, which we could discuss.

Regarding the vdom actually carrying the _events_ query, I’m cautious about it because it "couples" (for the lack of better wording) the view to the "intent" (query). However, I’d like to explore the concept, which might change my mind.

Thank you very much for your ideas and input. I’m definitely not closed to them, but I’d like some more talk to see how it may relieve our pain points without introducing new that are more severe. As you know, everything is a tradoff.

**Comment by milankinen** [comment]
>Particularly, transposition actually slowed it down, instead of, as we might think, speed it up. Perhaps we were doing it wrong and you might have some ideas on that, which we could discuss.

Yeah, transposition (or I'd like to say "observable lifting") is very expensive operation in it's current form. 

The good thing is that it can be implemented differently so that there is no need to propagate VDOM updates from left to root, instead one can perform diffing and patching in-place where the change occurs. The bad thing is that there are no existing VDOM implementations supporting this kind of feature, which means that the whole logic must be written from scratch. Like you said, it's a tradeoff.

>Regarding the vdom actually carrying the events query, I’m cautious about it because it "couples" the view to the "intent" (query).

I'm glad you noticed this! The very reason why `isolate` exists in the first place, is that "intent" (or its event sources) is not coupled to anything and because of this, events get mixed if no manual/forceful isolation is added. That's why I call it `IMV`. In other words, the view is "coupled" to the intent. 

In the end, either `V` or `I` must be coupled to the other one. What makes you cautious if it was `(M ->)V -> I` instead of `I -> (M ->)V`? What are the pros/cons in both approaches?

**Comment by axefrog** [comment]
Hi @milankinen, I don't think @TylorS would mind me mentioning that [I'm working on a project that addresses your concerns](https://github.com/frptools/salix#salix). I'm working on it full time, and progress is good. If you have any interest in discussing it further, let me know.

**Comment by milankinen** [comment]
Absolutely interested, @axefrog! Just ping me in Gitter?

**Comment by Frikki** [comment]
@milankinen 
> In the end, either V or I must be coupled to the other one. What makes you cautious if it was (M ->)V -> I instead of I -> (M ->)V? What are the pros/cons in both approaches?

I don’t know the pros and cons of each approach, as I haven’t investigated it in detail. What I can say is that when I design the architecture of applications, I make a clear separation between presentation, application logic, domain logic, and infrastructure. This means that my presentation layer is the only place where "intents" and views live. The model just transforms or combines data from the "intents" and is passed to view.

```typescript
// In the presentation layer
function SomeUIComponent(sources: Sources): Sinks {
  const { dom, sum$ } = sources;

  // Intent
  const btn = query(`.${SomeUIComponentStyle.btnClass}`, dom);

  const click$ = events('click', btn);

  // View
  const view$ = map(view, model$);

  return { view$, click$ };
}

function view(sum: Model): VNode {
  return div({}, [
    button(`.${SomeUIComponentStyle.btnClass}`, {}, `Click me`),
    h2(`Clicked ${sum} times!`),
  ]);
}

// In the application layer
function SumClicks(sources: Sources): Sinks {
  const { clicks$ } = sources;

  const sum$ = scan(sum, 0, constant(1, click$));

  // Here I may call my infrastructure to save values or push them to a third party consumer

  return { sum$ };
}
```

The result of the application service `SumClicks` is then by motorcycle looped into the presentation layer. In a way, the M has moved to the application layer, but it has become a data-flow component. The UI component just accepts some data, `sum$` in this case, and spits out some other data, `click$` in this case. It has no knowledge of the logic that the application performs. It does, however, has its own View Model (in lack of better words), which is only a data structure defining what data the View wants to render or act upon. As you see, the Intent is "loosly" coupled to the View. The only thing they share is the `btnClass`.

I am not sure what impact your suggestion would have on this architecture. Maybe you could throw out something?

**Comment by milankinen** [comment]
@Frikki, do you mean something like this?
```diff
// In the presentation layer
function SomeUIComponent(sources: Sources): Sinks {
  const { dom, sum$ } = sources;
+  // View
+  const view$ = map(view, model$);
  // Intent
-  const btn = query(`.${SomeUIComponentStyle.btnClass}`, dom);
+  const btn = query(`.${SomeUIComponentStyle.btnClass}`, view$);

  const click$ = events('click', btn);

-  // View
-  const view$ = map(view, model$);

  return { view$, click$ };
}

function view(sum: Model): VNode {
  return div({}, [
    button(`.${SomeUIComponentStyle.btnClass}`, {}, `Click me`),
    h2(`Clicked ${sum} times!`),
  ]);
}

// In the application layer
function SumClicks(sources: Sources): Sinks {
  const { clicks$ } = sources;

  const sum$ = scan(sum, 0, constant(1, click$));

  // Here I may call my infrastructure to save values or push them to a third party consumer

  return { sum$ };
}
```
I also advocate the clear separation of presentation and application logic and if I understood your concern right, I don't see that my proposal is in any contradiction to it.

**Comment by TylorS** [comment]
We actually have the facility for doing the querying one a `VNode` already built into `mostly-dom` just as an FYI. There are already, more limited of course, versions of `querySelector` and `querySelectorAll` functions that traverse a vNode. 

I'd worry a bit about the performance of this, as the query would have to happen much much more often than the current method's we have of hiding it away in the `DomSource`. Likely the query would have to happen every single time a view$ emits `n` number of times given `n` number of queries. Traversing `VNode`'s in as complex a way as `querySelector` and `querySelectorAll` is currently pretty slow on its own, too.

**Comment by milankinen** [comment]
In my opinion it really depends what we want to achieve with `querySelector`. `querySelector` returns an array of VDOM nodes, correct? If I recall right, `DomSource.select().elements()` returns an observable of DOM nodes. How about if `querySelector` worked in a similar way? 

Before I can say anything about the perf, I'd like to understand what is the real use case for `.elements()` (or `querySelector`) - is it used to get an access to the DOM nodes once, or get them every time when a change occurs? In the latter case, what changes should emit a new event: when selector nodes are added/removed? when vdom changes? when the selected DOM nodes really change? 

I've never found any use for `.elements()` so I can't answer to those questions... 😕  How do you use it?

(I intentionally skipped `.select().events()` because it's relatively easy to implement without any significant performance issues. Just a note: when I use `.select()`, I always combine it with `.events()`. But that's just me, hence I'd like to understand how other people are using those functions 😄)

**Comment by TylorS** [comment]
We've definitely used `elements()` in a few places. We've used it mostly for initial property/attribute/etc values, but also in a few other ways. 

> querySelector returns an array of VDOM nodes, correct? 

`querySelector` only returns a single VNode just like the DOM API of the same name. `querySeletorAll` will return `Array<VNode>`. Given that a `VNode` in your application always has an `.element` property attached it'd be easy enough to accomplish these things, even now. 

```typescript
import { querySelectorAll } from 'mostly-dom';
import { domEvent } from '@most/dom-event';

const view$: Stream<VNode> = map(view, model$);

const elements$ = elements('.foo', view$);

const click$ = events('.foo', 'click', view$);

function elements<T extends Element>(
  cssSelector: string, 
  view$: Stream<VNode>): Stream<Array<T>> 
{
  return view$
    .map(querySelectorAll(cssSelector))
    .map(vNodes => vNodes.map(vNode => vNode.element)));
}

function events<T extends Event>(
  cssSelector: string,
  eventType: string,
  view$: Stream<VNode>): Stream<Event>
{
  return view$
    .map(querySelectorAll(cssSelector))
    .map(vNodes => {
      return vNodes.reduce((event$, vNode) => {
        return merge(event$, domEvent(eventType, vNode.element))
      }, empty())
    })
}
```

**Comment by milankinen** [comment]
> Given that a VNode in your application always has an .element property attached it'd be easy enough to accomplish these things, even now.

Nice example! However, that's unfortunately not true because `querySelectorAll` is eager. Consider the following example:
```js
import {run} from '@motorcycle/run'
import {makeDomComponent, div, h1, button} from '@motorcycle/dom'
import * as O from "most"
import { querySelectorAll } from 'mostly-dom';
import { domEvent } from '@most/dom-event';

const DOM = makeDomComponent(document.querySelector("#app"))

function main({state$}) {
  const view$ = O.multicast(O.map(n => div([
    h1(n.toString()),
    button(".inc", ["++"])
  ]), state$))
  
  const click$ = events(".inc", "click", view$)
  return {
     // swap these lines and the app will work
     action$: O.constant(1, click$),
     view$
  }
}

function effects(sinks) {
  return {
     state$: O.scan((x, y) => x + y, 0, sinks.action$),
     DOM: DOM(sinks)
  }
}

function events(cssSelector, eventType, view$) {
  return view$
    .map(querySelectorAll(cssSelector))
    .map(vNodes => {
      console.log(vNodes.map(n => n.element))
      return vNodes.reduce((event$, vNode) => {
        // ERROR: vNode.element === undefined
        return O.merge(event$, domEvent(eventType, vNode.element))
      }, O.empty())
    })
    .switch()
}

run(main, effects)
``` 
Now the `events()` subscription gets evaluated first, which means that vnodes are not mounted yet (= no `.element`), resulting in an error. If we added a `!== undefined` check, it would prevent the error but then we wouldn't be able to access the elements after the mount.

**Comment by milankinen** [comment]
>We've definitely used `elements()` in a few places. We've used it mostly for initial property/attribute/etc values, but also in a few other ways.

Interesting. What made you to choose `element` instead of hooks? How about the "other ways"? Do you have any examples what was the use case?

**Comment by TylorS** [comment]
Ahhh yes of course, the querySelector stuff would never really work as is. If an events module were added to mostly-dom it could instead "attach" listeners to the vnodes that begin listening upon patch() instead. I suppose something similar could be done with elements as well, which could then allow this approach along with the current. 

Does that seem sufficient?

**Comment by milankinen** [comment]
>Ahhh yes of course, the querySelector stuff would never really work as is. If an events module were added to mostly-dom it could instead "attach" listeners to the vnodes that begin listening upon patch() instead. I suppose something similar could be done with elements as well, which could then allow this approach along with the current.

Exactly! 👍 And I still want to emphasize that for the event listening case, there isn't even need to monitor any `patch` invocations -- instead, you can use single event listener per event name at the top of your DOM tree and delegate events to their corresponding observers (and I really mean **single listener**, not like e.g. in `@cycle/dom` which uses event listener per isolation scope root and spoiling the performance benefits when handling lists)

**Comment by TylorS** [comment]
I've been thinking about this for a little while, and I'm going to implement this in `mostly-dom` when I have the time, and will open a PR and reference back here. 

The way I've been envisioning `mostly-dom` is finally starting to take shape, as I just got specific typings for each individual hyperscript-helper function merged and released in so that only the properties that correspond to each element is accepted. That took a lot of work, but now it should be possible to take the type parameters and use them to create typings for only the type of events that are available for a given element as well!

**Comment by TylorS** [comment]
`mostly-dom` has been updated to v3.4.0 with the ability to add event handlers like so

```typescript
import { div, VNodeEvent, ElementVNode  } from 'mostly-dom'

div({
  on: { click: clickHandler }
})

function clickHandler (
  event: VNodeEvent<HTMLDivElement, MouseEvent>,
  vNode: ElementVNode<HTMLDivElement>): void
{
  const { target } = event // VNodeEvent is used to ensure `target` to be of a type specific element
}
```

So a function that returns a stream of events from a VNode could now be developed using the `querySelector`/`querySelectorAll` and the new event listeners module. I think I'm in favor of it existing in an external package though, as it's not something I see myself using frequently as I like the current style of usage. 

In regards to transposition, I'm not in favor of reintroducing that behavior. I found it to be too magical and led to very hard to maintain code at my last job (we used Cycle pre-diversity with Rx 4). In my current work, we've introduced many most.js helpers that make dealing with child components much nicer. Those helpers will be introduced as a package, likely named `@motorcycle/stream`, when it is time to rebuild on top of `@most/core` (the next version of most in development) to help streamline motorcycle development even further.

@milankinen I'll leave it up to you if you feel this issue has been addressed or to have it continue!

**Comment by TylorS** [comment]
As a quick example https://www.webpackbin.com/bins/-KmCa3g9Z3xzmRlcpnap

**Comment by milankinen** [comment]
>... I think I'm in favor of it existing in an external package though ...

Yeah, an external package sounds the best option in this case. 👍 From my behalf, this issue can be closed. I'll probably do some personal PoC'ing of different options and be back in touch if I managed to find something cool. 😄 

>In regards to transposition, I'm not in favor of reintroducing that behavior. I found it to be too magical and led to very hard to maintain code at my last job

What made it feel "too magical"? How about if you compare these two code snippets with similar behaviour?

```js
const vdom$ = doPipe(
  [children.dom$, title$, value$],
  O.combine((childVNodes, title, value) =>
   div([
     h1([title]),
     input({type: 'text', value}),
     div([value]),
     ul(childVNodes)
   ])))
// vdom$ :: Stream<VNode>
```
Versus

```js
const vdom$ = DOM.combine(div([
  h1([title$]),
  input({type: 'text', value: value$}),
  div([value]),
  ul(children.dom$)
]))
// vdom$ :: Stream<VNode>
```

Do you feel that the other is "more magical" than the another one? If so, what makes you feel like it?

**Comment by TylorS** [comment]
The way that I write all of my components is using my helper from @mostjs-community `combineObj`

```typescript
type Model = {
  value: string,
  title: string,
  children: Array<VNode>
}

const model$ = combineObj({
  children$,
  title$,
  value$
})

const view$ = map(view, model$)

function view(model: Model) {
  const { children, title, value } = model 

  return div(...)
}
```

I, of course, find this example to be very clear and understandable, and if you could see my code at work almost all of it follows this pattern, and refactoring and adding features has been easy even for the very junior team member we have.

At my previous job we made heavy use of transposition, and I remember when going to fix a bug or add a new feature everything started to become a lot more fragile when it came to refactoring, that may sound vague and it's not intentionally but it's hard to remember specifics at this point. My own experience however was quite negative.

I also prefer, and feel it's a better practice to work as much without streams as possible. For instance I can statically test my view in any number if ways just by passing a different model and inspecting the results. There is however no simple way of testing it without handling the transposition inside the view as in your example. 

Lastly there comes the aspects regarding performance, and they could definitely be felt in older cycle versions that had the option. Without writing a virtual-dom library that handles transposition natively I don't see it ever being very fast. If you remember it was a giant recursive `combine` that forces creating streams where there were none just to handle where there might be some.

**Comment by TylorS** [comment]
@milankinen Here's a quick POC of the event stuff that could potentially be done with `mostly-dom`

**Comment by milankinen** [comment]
>At my previous job we made heavy use of transposition, and I remember when going to fix a bug or add a new feature everything started to become a lot more fragile when it came to refactoring, that may sound vague and it's not intentionally but it's hard to remember specifics at this point. My own experience however was quite negative.

Damn, that's sad to hear. 😕  I guess it's very subjective then. 😄 I've always found transposition very natural and concise way of declaring the UI: virtual dom is just a JSON object, so `DOM.combine(vdom)` is just a `combine*` variant like `@most/combineObj`.

>I also prefer, and feel it's a better practice to work as much without streams as possible.
+1

>Without writing a virtual-dom library that handles transposition natively I don't see it ever being very fast.

Yes, the implementation has to support transposition natively in order to make it fast. But if the implementation is done correctly, single updates are **extremely fast** because they can be performed in-place.

>Here's a quick POC of the event stuff that could potentially be done with mostly-dom
https://www.webpackbin.com/bins/-KnFP2QXoDkCtX-nttP7

Nice! That's very close to my original proposal. Actually, if you refactor the event listener attachment stuff into its own `events(view$, selector, type)` function, you can call that from you "component" and eliminate the needed isolation completely, e.g. like this https://www.webpackbin.com/bins/-KnQQKFqVsEMWvS79nFl 😊 

(The example is not working atm, I suspect that `mostly-dom` has some kind of bug for event listener updates because the first `props.on.click` gets called even if the vdom has been patched with newer version of it).

**Comment by TylorS** [comment]
If you found a bug, feel free to open an issue and we can get it fixed.
